linera-service/src/cli_wrappers/local_net.rs

```diff
@@ -138,6 +142,82 @@ where
     }
 }
 
+/// A data structure to store the current state of available ports.
+#[cfg(any(test, feature = "test"))]
+struct PortShifts {
+    semaphore_ports: Arc<Semaphore>,
+    index: Arc<RwLock<usize>>,
+}
+
+/// The guard for the shift in ports.
+struct IndexPortGuard {
+    _semaphore_guard: OwnedSemaphorePermit,
+}
+
+#[cfg(any(test, feature = "test"))]
+impl PortShifts {
+    pub fn new(n_ports: usize) -> Self {
+        let semaphore_ports = Arc::new(Semaphore::new(n_ports));
```

Comment:
Do we need a semaphore? (which I understand is used to limit the number of active ports?)

Reply:
Yes, the semaphore is used for that purpose.
I think it is useful. If we set it to 1 then we restore the `INTEGRATION_TEST_GUARD` and that is useful.
Also, each run typically involved 4 validators and 4 shards so 16 connections. Limiting this is I think a prudent choice.